core/templates/pages/translate-custom-parser.ts

```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Custom parser for translations.
+ */
+
+import { TranslateDefaultParser, TranslateParser } from '@ngx-translate/core';
+import constants from 'assets/constants';
+import MessageFormat from 'messageformat';
+import { I18nLanguageCodeService } from 'services/i18n-language-code.service';
+
+export class TranslateCustomParser extends TranslateParser {
+  messageFormat;
+  constructor(
+    private translateDefaultParser: TranslateDefaultParser,
+    private i18nLanguageCodeService: I18nLanguageCodeService
+  ) {
+    super();
+    this.translateDefaultParser.templateMatcher = /<\[\s?([^{}\s]*)\s?\]>/g;
+    this.messageFormat = new MessageFormat();
+  }
+
+  interpolate(
+      expr: string | Function,
+      params?: { [key: string]: number | string | boolean }): string {
+    let interpolate = this.translateDefaultParser.interpolate(expr, params);
+    if (params) {
+      if (params.hasOwnProperty('plural') && interpolate !== undefined) {
+        if (params.plural) {
+          try {
+            interpolate = this.messageFormat.compile(
+              interpolate, this.i18nLanguageCodeService
+                .getCurrentI18nLanguageCode())(params);
```

Comment:
here the language used by the default parser and the language sent in the second parameter should be the same.
But we cannot tell that whether default parser used the primary language translations or the backup language translations to interpolate the expression. So, here I first sent the interpolated string by default parser and the primary language code. If these mismatch, then message format will throw error in which case, the default language is used.

Reply:
fixed